server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftScanClientHandler.java

```diff
@@ -0,0 +1,531 @@
+      throws ThriftSecurityException {
+    try {
+      log.trace("Got {} message from user: {}", request, credentials.getPrincipal());
+      if (!security.canPerformSystemActions(credentials)) {
+        log.warn("Got {} message from user: {}", request, credentials.getPrincipal());
+        throw new ThriftSecurityException(credentials.getPrincipal(),
+            SecurityErrorCode.PERMISSION_DENIED);
+      }
+    } catch (ThriftSecurityException e) {
+      log.warn("Got {} message from unauthenticatable user: {}", request, e.getUser());
+      if (context.getCredentials().getToken().getClass().getName()
+          .equals(credentials.getTokenClassName())) {
+        log.error("Got message from a service with a mismatched configuration."
+            + " Please ensure a compatible configuration.", e);
+      }
+      throw e;
+    }
+
+    if (server.getLock() == null || !server.getLock().wasLockAcquired()) {
+      log.debug("Got {} message before my lock was acquired, ignoring...", request);
+      throw new RuntimeException("Lock not acquired");
+    }
+
+    if (server.getLock() != null && server.getLock().wasLockAcquired()
+        && !server.getLock().isLocked()) {
+      Halt.halt(1, () -> {
+        log.info("Tablet server no longer holds lock during checkPermission() : {}, exiting",
+            request);
+        server.gcLogger.logGCInfo(server.getConfiguration());
+      });
+    }
+
+    if (lock != null) {
+      ZooUtil.LockID lid =
+          new ZooUtil.LockID(context.getZooKeeperRoot() + Constants.ZMANAGER_LOCK, lock);
+
+      try {
+        if (!ServiceLock.isLockHeld(server.managerLockCache, lid)) {
+          // maybe the cache is out of date and a new manager holds the
+          // lock?
+          server.managerLockCache.clear();
+          if (!ServiceLock.isLockHeld(server.managerLockCache, lid)) {
+            log.warn("Got {} message from a manager that does not hold the current lock {}",
+                request, lock);
+            throw new RuntimeException("bad manager lock");
+          }
+        }
+      } catch (Exception e) {
+        throw new RuntimeException("bad manager lock", e);
+      }
+    }
+  }
+
+  private NamespaceId getNamespaceId(TCredentials credentials, TableId tableId)
+      throws ThriftSecurityException {
+    try {
+      return server.getContext().getNamespaceId(tableId);
+    } catch (TableNotFoundException e1) {
+      throw new ThriftSecurityException(credentials.getPrincipal(),
+          SecurityErrorCode.TABLE_DOESNT_EXIST);
+    }
+  }
+
+  private ScanDispatcher getScanDispatcher(KeyExtent extent) {
+    if (extent.isRootTablet() || extent.isMeta()) {
+      // dispatcher is only for user tables
+      return null;
+    }
+
+    return context.getTableConfiguration(extent.tableId()).getScanDispatcher();
+  }
+
+  @Override
+  public InitialScan startScan(TInfo tinfo, TCredentials credentials, TKeyExtent textent,
```

Comment:
Were all of these overridden methods copied exactly from TabletClientHandler?  Wondering if there any changes to look at.

Reply:
Should be an exact copy, there were no changes to the Thrift method signatures, just moving them. IIRC, I copied the code and pasted it into a new class.